src/main/java/org/folio/service/transactions/batch/BatchEncumbranceService.java

```diff
@@ -0,0 +1,113 @@
+      budget.setEncumbered(newEncumbered);
+    });
+    calculateBudgetSummaryFields(budget);
+    updateBudgetMetadata(budget, encumbrances.get(0));
+  }
+
+  private void updateBudgetForEncumbranceUpdate(Budget budget, List<Transaction> encumbrances,
+      Map<String, Transaction> existingTransactions) {
+    CurrencyUnit currency = Monetary.getCurrency(encumbrances.get(0).getCurrency());
+    encumbrances.forEach(encumbrance -> {
+      Transaction existingEncumbrance = existingTransactions.get(encumbrance.getId());
+      if (existingEncumbrance != null && isNotFromReleasedExceptToUnreleased(encumbrance, existingEncumbrance)) {
+        updateBudget(budget, currency, encumbrance, existingEncumbrance);
+      }
+    });
+    calculateBudgetSummaryFields(budget);
+    updateBudgetMetadata(budget, encumbrances.get(0));
+  }
+
+  private void updateBudget(Budget budget, CurrencyUnit currency, Transaction encumbrance, Transaction existingEncumbrance) {
+    double newEncumbered = budget.getEncumbered();
+    if (isEncumbranceReleased(encumbrance)) {
+      newEncumbered = subtractMoney(newEncumbered, existingEncumbrance.getAmount(), currency);
+      encumbrance.setAmount(0d);
+    } else  if (isTransitionFromUnreleasedToPending(encumbrance, existingEncumbrance)) {
+      encumbrance.setAmount(0d);
+      encumbrance.getEncumbrance().setInitialAmountEncumbered(0d);
+      newEncumbered = subtractMoney(newEncumbered, existingEncumbrance.getAmount(), currency);
+    } else if (isTransitionFromPendingToUnreleased(encumbrance, existingEncumbrance)) {
+      double newAmount = subtractMoney(encumbrance.getEncumbrance().getInitialAmountEncumbered(),
+        existingEncumbrance.getEncumbrance().getAmountAwaitingPayment(), currency);
+      newAmount = subtractMoney(newAmount, existingEncumbrance.getEncumbrance().getAmountExpended(), currency);
+      encumbrance.setAmount(newAmount);
+      newEncumbered = sumMoney(currency, newEncumbered, newAmount);
+    } else if (isTransitionFromReleasedToUnreleased(encumbrance, existingEncumbrance)) {
+      double newAmount = subtractMoney(encumbrance.getEncumbrance().getInitialAmountEncumbered(),
+        encumbrance.getEncumbrance().getAmountAwaitingPayment(), currency);
+      newAmount = subtractMoney(newAmount, encumbrance.getEncumbrance().getAmountExpended(), currency);
+      encumbrance.setAmount(newAmount);
+      newEncumbered = sumMoney(newEncumbered, newAmount, currency);
+    } else {
+      newEncumbered = sumMoney(newEncumbered, encumbrance.getAmount(), currency);
+      newEncumbered = subtractMoney(newEncumbered, existingEncumbrance.getAmount(), currency);
+    }
+    budget.setEncumbered(newEncumbered);
+  }
+
+
+  private boolean isEncumbranceReleased(Transaction encumbrance) {
+    return encumbrance.getEncumbrance().getStatus() == Encumbrance.Status.RELEASED;
+  }
+
+  private boolean isNotFromReleasedExceptToUnreleased(Transaction newEncumbrance, Transaction existingEncumbrance) {
```

Comment:
maybe better to rename to have the same convention?
isTransitionNotFromReleasedToUnreleased

Reply:
As I understand it, "`isTransitionNotFromReleasedToUnreleased`" would have a different logic:
`!(existingEncumbrance.getEncumbrance().getStatus() == Encumbrance.Status.RELEASED && newEncumbrance.getEncumbrance().getStatus() == Encumbrance.Status.UNRELEASED)`
which could also be written as
`existingEncumbrance.getEncumbrance().getStatus() != Encumbrance.Status.RELEASED || newEncumbrance.getEncumbrance().getStatus() != Encumbrance.Status.UNRELEASED`
Here we want to prevent a change on a released encumbrance, except if the encumbrance is going to be released (I imagine initially we never wanted to change released encumbrances but we added that exception at some point).
To be fair I am not sure we still need this logic, but I kept it as I try to keep logic close to the previous implementation.